### PR TITLE
Close #3

### LINKED ISSUE #3: Add tags manage
1. create tag
2. get tags
3. update tag
4. delete tag
5. get by tag
6. batch tagging
7. batch untagging
8. get openid list by tag
9. get blacklist
10. batch blacklist
11. batch unblacklist

### PR DESCRIPTION
**Pull request by jalog**
tags manage

### CODE REVIEW

**Comment by calidion** [comment]
test cases for tags needed.

File: lib/tags.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    var next_openid = openid || '';
+    send(settings, app, 
+      'https://api.weixin.qq.com/cgi-bin/user/tag/get?', {
+        tagid: tagId,
+        next_openid: next_openid
+      }, 
+      cb 
+    );
+  },
+  tagUsers: function(settings, app, tagId, openid_list, cb) {
```
Comment (calidion) [inline lib/tags.js:57]: openid_list to OpenIdList

File: lib/tags.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  getblacklist: function(settings, app, openid, cb) {
+    var params = {};
+    if (openid) {
+      params = {
+        next_openid: openid
+      };
+    }
+    get(settings, app, 'https://api.weixin.qq.com/cgi-bin/tags/members/getblacklist?', params, cb);
+  },
+  batchblacklist: function(settings, app, openid_list, cb) {
```
Comment (calidion) [inline lib/tags.js:92]: to openIds or openIdList
to batchBlackList

File: lib/tags.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    post(settings, app, 
+      'https://api.weixin.qq.com/cgi-bin/tags/members/getblacklist?', {
+        openid_list: openid_list
+      }, 
+      cb
+    );
+  },
+  batchunblacklist: function(settings, app, openid_list, cb) {
```
Comment (calidion) [inline lib/tags.js:100]: to openIds or openIdList
to batchUnblackList

File: lib/tags.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  untagUsers: function(settings, app, tagId, openid_list, cb) {
+    send(settings, app,
+      'https://api.weixin.qq.com/cgi-bin/tags/members/batchuntagging?', {
+        openid_list: openid_list,
+        tagid: tagId
+      },
+      cb
+    );
+  },
+  getidlist: function(settings, app, openid, cb) {
```
Comment (calidion) [inline lib/tags.js:75]: getidlist to getIdList

File: lib/tags.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    send(settings, app,
+      'https://api.weixin.qq.com/cgi-bin/tags/getidlist?', {
+        openid: openid
+      },
+      cb
+    );
+  },
+  getblacklist: function(settings, app, openid, cb) {
```
Comment (calidion) [inline lib/tags.js:83]: to getBlackList

File: lib/tags.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    send(settings, app,
+      'https://api.weixin.qq.com/cgi-bin/tags/delete?', {
+        tag: {
+          id: tagId
+        }
+      },
+      cb
+    );
+  },
+  list: function(settings, app, tagId, openid, cb) {
```
Comment (calidion) [inline lib/tags.js:47]: better use openId

**Review by calidion — COMMENTED** [review]
some fixes before merging.

File: lib/tags.js (hunk 1 above)
Comment (jalog) [reply lib/tags.js:57]: OK

**Comment by jalog** [comment]
> test cases for tags needed.

OK, I've added test cases.

File: lib/tags.js (hunk 6 above)
Comment (jalog) [reply lib/tags.js]: fixed.

File: lib/tags.js (hunk 5 above)
Comment (jalog) [reply lib/tags.js]: fixed.

File: lib/tags.js (hunk 4 above)
Comment (jalog) [reply lib/tags.js]: fixed.

File: lib/tags.js (hunk 3 above)
Comment (jalog) [reply lib/tags.js]: fixed.

File: lib/tags.js (hunk 2 above)
Comment (jalog) [reply lib/tags.js]: fixed.

**Pull request merged**